File: backend/routing.py
"""Route fetching with OSRM public demo — now with turn-by-turn steps.
Graceful interpolated fallback so the demo never hard-fails."""
from __future__ import annotations
import math
from typing import List, Tuple, Dict
import httpx

OSRM_URL = "https://router.project-osrm.org/route/v1/driving"

MANEUVER_ICONS = {
    "turn": "turn", "new name": "straight", "depart": "depart",
    "arrive": "arrive", "merge": "merge", "on ramp": "ramp",
    "off ramp": "ramp", "fork": "fork", "end of road": "end",
    "continue": "straight", "roundabout": "roundabout",
    "rotary": "roundabout", "roundabout turn": "roundabout",
    "notification": "info", "exit roundabout": "roundabout",
    "exit rotary": "roundabout",
}

def _build_instruction(step: Dict) -> str:
    """Build a human-readable instruction from OSRM step."""
    maneuver = step.get("maneuver", {})
    mtype = maneuver.get("type", "")
    modifier = maneuver.get("modifier", "")
    name = step.get("name", "")
    dist = step.get("distance", 0)

    if mtype == "depart":
        return f"Head {modifier} on {name}" if name else f"Head {modifier}"
    if mtype == "arrive":
        return "You have arrived at your destination"
    if mtype == "turn":
        direction = modifier.replace(" ", " ").title()
        return f"Turn {direction} onto {name}" if name else f"Turn {direction}"
    if mtype == "roundabout" or mtype == "rotary":
        exit_num = maneuver.get("exit", "")
        return f"Take exit {exit_num} at the roundabout onto {name}" if name else f"Take exit {exit_num} at the roundabout"
    if mtype == "merge":
        return f"Merge {modifier} onto {name}" if name else f"Merge {modifier}"
    if mtype == "fork":
        return f"Keep {modifier} onto {name}" if name else f"Keep {modifier}"
    if mtype == "new name" or mtype == "continue":
        return f"Continue onto {name}" if name else "Continue straight"
    if mtype in ("on ramp", "off ramp"):
        return f"Take the ramp onto {name}" if name else "Take the ramp"
    if mtype == "end of road":
        return f"Turn {modifier} at the end of the road onto {name}" if name else f"Turn {modifier}"

    return f"{mtype.title()} {modifier} {name}".strip()


async def fetch_routes(
    from_lat: float, from_lng: float,
    to_lat: float, to_lng: float,
    alternatives: int = 2,
) -> List[Dict]:
    """Returns routes with polylines AND turn-by-turn navigation steps."""
    coords = f"{from_lng},{from_lat};{to_lng},{to_lat}"
    url = f"{OSRM_URL}/{coords}"
    params = {
        "overview": "full",
        "geometries": "geojson",
        "alternatives": "true" if alternatives > 1 else "false",
        "steps": "true",  # Enable turn-by-turn
    }
    try:
        async with httpx.AsyncClient(timeout=8.0) as client:
            r = await client.get(url, params=params)
            r.raise_for_status()
            data = r.json()
        routes = []
        for i, rt in enumerate(data.get("routes", [])):
            coords_list = rt["geometry"]["coordinates"]  # [lng, lat]
            polyline = [(c[1], c[0]) for c in coords_list]

            # Extract turn-by-turn steps
            steps = []
            for leg in rt.get("legs", []):
                for step in leg.get("steps", []):
                    maneuver = step.get("maneuver", {})
                    step_coords = step.get("geometry", {}).get("coordinates", [])
                    step_polyline = [(c[1], c[0]) for c in step_coords] if step_coords else []

                    steps.append({
                        "instruction": _build_instruction(step),
                        "distance_m": round(step.get("distance", 0)),
                        "duration_s": round(step.get("duration", 0)),
                        "name": step.get("name", ""),
                        "maneuver_type": maneuver.get("type", ""),
                        "modifier": maneuver.get("modifier", ""),
                        "icon": MANEUVER_ICONS.get(maneuver.get("type", ""), "straight"),
                        "location": [maneuver.get("location", [0, 0])[1], maneuver.get("location", [0, 0])[0]],  # [lat, lng]
                        "polyline": step_polyline,
                    })

            routes.append({
                "id": f"r{i}",
                "polyline": polyline,
                "distance_m": rt["distance"],
                "duration_s": rt["duration"],
                "source": "osrm",
                "steps": steps,
            })
        if routes:
            return routes
    except Exception as e:
        print(f"[routing] OSRM failed: {e}, using fallback")

    return [_fallback_route(from_lat, from_lng, to_lat, to_lng)]


def _fallback_route(a_lat: float, a_lng: float, b_lat: float, b_lng: float) -> Dict:
    N = 50
    poly = []
    mid_lat = (a_lat + b_lat) / 2
    mid_lng = (a_lng + b_lng) / 2
    dx = b_lng - a_lng
    dy = b_lat - a_lat
    perp_lat = -dx * 0.05
    perp_lng = dy * 0.05
    bend_lat = mid_lat + perp_lat
    bend_lng = mid_lng + perp_lng

    def bezier(t):
        u = 1 - t
        lat = u * u * a_lat + 2 * u * t * bend_lat + t * t * b_lat
        lng = u * u * a_lng + 2 * u * t * bend_lng + t * t * b_lng
        return (lat, lng)

    for i in range(N + 1):
        poly.append(bezier(i / N))

    dist = 0.0
    for i in range(1, len(poly)):
        from risk_engine import haversine_m
        dist += haversine_m(poly[i - 1][0], poly[i - 1][1], poly[i][0], poly[i][1])

    # Generate fallback steps
    steps = [
        {"instruction": "Head toward your destination", "distance_m": round(dist),
         "duration_s": round(dist / 11.0), "name": "", "maneuver_type": "depart",
         "modifier": "straight", "icon": "depart", "location": list(poly[0]), "polyline": poly[:25]},
        {"instruction": "You have arrived at your destination", "distance_m": 0,
         "duration_s": 0, "name": "", "maneuver_type": "arrive",
         "modifier": "", "icon": "arrive", "location": list(poly[-1]), "polyline": poly[25:]},
    ]

    return {
        "id": "r0",
        "polyline": poly,
        "distance_m": dist,
        "duration_s": dist / 11.0,
        "source": "fallback",
        "steps": steps,
    }

File: web/lib/backend/routing.ts
/**
 * Route fetching with OSRM public demo -- now with turn-by-turn steps.
 * Graceful interpolated fallback so the demo never hard-fails.
 */

export type Polyline = [number, number][];

export interface RouteStep {
  instruction: string;
  distance_m: number;
  duration_s: number;
  name: string;
  maneuver_type: string;
  modifier: string;
  icon: string;
  location: [number, number];
  polyline: Polyline;
}

export interface Route {
  id: string;
  polyline: Polyline;
  distance_m: number;
  duration_s: number;
  source: string;
  steps: RouteStep[];
}

const OSRM_URL = "https://router.project-osrm.org/route/v1/driving";

export const MANEUVER_ICONS: Record<string, string> = {
  "turn": "turn",
  "new name": "straight",
  "depart": "depart",
  "arrive": "arrive",
  "merge": "merge",
  "on ramp": "ramp",
  "off ramp": "ramp",
  "fork": "fork",
  "end of road": "end",
  "continue": "straight",
  "roundabout": "roundabout",
  "rotary": "roundabout",
  "roundabout turn": "roundabout",
  "notification": "info",
  "exit roundabout": "roundabout",
  "exit rotary": "roundabout",
};

function buildInstruction(step: Record<string, unknown>): string {
  const maneuver = (step.maneuver || {}) as Record<string, unknown>;
  const mtype = (maneuver.type || "") as string;
  const modifier = (maneuver.modifier || "") as string;
  const name = (step.name || "") as string;

  if (mtype === "depart") {
    return name ? `Head ${modifier} on ${name}` : `Head ${modifier}`;
  }
  if (mtype === "arrive") {
    return "You have arrived at your destination";
  }
  if (mtype === "turn") {
    const direction = modifier
      .split(" ")
      .map((w) => w.charAt(0).toUpperCase() + w.slice(1))
      .join(" ");
    return name ? `Turn ${direction} onto ${name}` : `Turn ${direction}`;
  }
  if (mtype === "roundabout" || mtype === "rotary") {
    const exitNum = maneuver.exit || "";
    return name
      ? `Take exit ${exitNum} at the roundabout onto ${name}`
      : `Take exit ${exitNum} at the roundabout`;
  }
  if (mtype === "merge") {
    return name ? `Merge ${modifier} onto ${name}` : `Merge ${modifier}`;
  }
  if (mtype === "fork") {
    return name ? `Keep ${modifier} onto ${name}` : `Keep ${modifier}`;
  }
  if (mtype === "new name" || mtype === "continue") {
    return name ? `Continue onto ${name}` : "Continue straight";
  }
  if (mtype === "on ramp" || mtype === "off ramp") {
    return name ? `Take the ramp onto ${name}` : "Take the ramp";
  }
  if (mtype === "end of road") {
    return name
      ? `Turn ${modifier} at the end of the road onto ${name}`
      : `Turn ${modifier}`;
  }

  return `${mtype.charAt(0).toUpperCase() + mtype.slice(1)} ${modifier} ${name}`.trim();
}

const EARTH_R = 6371000.0;

function haversineM(
  aLat: number,
  aLng: number,
  bLat: number,
  bLng: number,
): number {
  const p1 = (aLat * Math.PI) / 180;
  const p2 = (bLat * Math.PI) / 180;
  const dp = ((bLat - aLat) * Math.PI) / 180;
  const dl = ((bLng - aLng) * Math.PI) / 180;
  const h =
    Math.sin(dp / 2) ** 2 +
    Math.cos(p1) * Math.cos(p2) * Math.sin(dl / 2) ** 2;
  return 2 * EARTH_R * Math.asin(Math.sqrt(h));
}

function fallbackRoute(
  aLat: number,
  aLng: number,
  bLat: number,
  bLng: number,
): Route {
  const N = 50;
  const midLat = (aLat + bLat) / 2;
  const midLng = (aLng + bLng) / 2;
  const dx = bLng - aLng;
  const dy = bLat - aLat;
  const perpLat = -dx * 0.05;
  const perpLng = dy * 0.05;
  const bendLat = midLat + perpLat;
  const bendLng = midLng + perpLng;

  function bezier(t: number): [number, number] {
    const u = 1 - t;
    const lat = u * u * aLat + 2 * u * t * bendLat + t * t * bLat;
    const lng = u * u * aLng + 2 * u * t * bendLng + t * t * bLng;
    return [lat, lng];
  }

  const poly: Polyline = [];
  for (let i = 0; i <= N; i++) {
    poly.push(bezier(i / N));
  }

  let dist = 0.0;
  for (let i = 1; i < poly.length; i++) {
    dist += haversineM(
      poly[i - 1][0],
      poly[i - 1][1],
      poly[i][0],
      poly[i][1],
    );
  }

  const steps: RouteStep[] = [
    {
      instruction: "Head toward your destination",
      distance_m: Math.round(dist),
      duration_s: Math.round(dist / 11.0),
      name: "",
      maneuver_type: "depart",
      modifier: "straight",
      icon: "depart",
      location: [poly[0][0], poly[0][1]],
      polyline: poly.slice(0, 25),
    },
    {
      instruction: "You have arrived at your destination",
      distance_m: 0,
      duration_s: 0,
      name: "",
      maneuver_type: "arrive",
      modifier: "",
      icon: "arrive",
      location: [poly[poly.length - 1][0], poly[poly.length - 1][1]],
      polyline: poly.slice(25),
    },
  ];

  return {
    id: "r0",
    polyline: poly,
    distance_m: dist,
    duration_s: dist / 11.0,
    source: "fallback",
    steps,
  };
}

export async function fetchRoutes(
  fromLat: number,
  fromLng: number,
  toLat: number,
  toLng: number,
  alternatives: number = 2,
): Promise<Route[]> {
  const coords = `${fromLng},${fromLat};${toLng},${toLat}`;
  const url = `${OSRM_URL}/${coords}`;
  const params = new URLSearchParams({
    overview: "full",
    geometries: "geojson",
    alternatives: alternatives > 1 ? "true" : "false",
    steps: "true",
  });

  try {
    const controller = new AbortController();
    const timeoutId = setTimeout(() => controller.abort(), 8000);

    const r = await fetch(`${url}?${params.toString()}`, {
      signal: controller.signal,
    });
    clearTimeout(timeoutId);

    if (!r.ok) throw new Error(`HTTP ${r.status}`);
    const data = await r.json();

    const routes: Route[] = [];
    for (let i = 0; i < (data.routes || []).length; i++) {
      const rt = data.routes[i];
      const coordsList: [number, number][] = rt.geometry.coordinates;
      const polyline: Polyline = coordsList.map(
        (c: [number, number]) => [c[1], c[0]] as [number, number],
      );

      // Extract turn-by-turn steps
      const steps: RouteStep[] = [];
      for (const leg of rt.legs || []) {
        for (const step of leg.steps || []) {
          const maneuver = step.maneuver || {};
          const stepCoords: [number, number][] = (step.geometry
            ?.coordinates || []) as [number, number][];
          const stepPolyline: Polyline = stepCoords.map(
            (c: [number, number]) => [c[1], c[0]] as [number, number],
          );

          const loc = maneuver.location || [0, 0];
          steps.push({
            instruction: buildInstruction(step),
            distance_m: Math.round(step.distance || 0),
            duration_s: Math.round(step.duration || 0),
            name: step.name || "",
            maneuver_type: maneuver.type || "",
            modifier: maneuver.modifier || "",
            icon: MANEUVER_ICONS[maneuver.type || ""] || "straight",
            location: [loc[1], loc[0]],
            polyline: stepPolyline,
          });
        }
      }

      routes.push({
        id: `r${i}`,
        polyline,
        distance_m: rt.distance,
        duration_s: rt.duration,
        source: "osrm",
        steps,
      });
    }
    if (routes.length > 0) return routes;
  } catch (e) {
    console.log(`[routing] OSRM failed: ${e}, using fallback`);
  }

  return [fallbackRoute(fromLat, fromLng, toLat, toLng)];
}
